Watcher V2 with smart contract

koirikivi opened the pull request:
This PR includes the first version of the V2 smart contract version of the Watcher. The main change is that there's now a smart contract, and arbitrage and liquidation are done using funds deposited to that contract.

The smart contracts and related utilities are in the `smart-contracts` directory.

The main contract is `Watcher.sol`. It has two roles, `OWNER`, which can deposit and withdraw funds
and grant other roles, and `EXECUTOR` which can execute transactions.

Please see `smart-contracts/README.md` for more instructions.

koirikivi committed:
Smartcontracts package

koirikivi committed:
Hardhat basis

koirikivi committed:
Rename Greeter to Watcher

koirikivi committed:
smart-contracts: typescript, working smoke test

koirikivi committed:
smart-contracts: make Watcher ownable, openzeppelin dep, package-lock

koirikivi committed:
IERC20 from openzeppelin

koirikivi committed:
smart-contracts: npm test

koirikivi committed:
smart-contracts: Check fixed amount arbitrage

koirikivi committed:
Refactoring

koirikivi committed:
Watcher: abritrage

koirikivi committed:
small refactoring

koirikivi committed:
Working arbitrage function and tests

koirikivi committed:
Test WRBTC wrapping

koirikivi committed:
Refactoring, Waffle goodness

koirikivi committed:
fix/test WRBTC as target token

koirikivi committed:
quick and dirty v2 backend basis

koirikivi committed:
Remove unnecessary stuff

koirikivi committed:
Quick and dirty Arbitrage V2 backend integration

koirikivi committed:
Whoops, fix minReturn in Watcher.sol

koirikivi committed:
Basis for liquidation v2

koirikivi committed:
quickfix failing arbitrage tests after sending away rbtc

koirikivi committed:
Refactoring

koirikivi committed:
Watcher v2 liquidation basis

koirikivi committed:
Token deposit, withdrawals

koirikivi committed:
Remove unnecessary comment

koirikivi committed:
Test RBTC withdrawal too

koirikivi committed:
Convert to solidity 0.8

koirikivi committed:
Replace Ownable with AccessControl

koirikivi committed:
V2 liquidator backend/tests

koirikivi committed:
Remove unnecessary OpenZeppelin contracts

koirikivi committed:
Update watcher abi

koirikivi committed:
Solidity style guide changes

koirikivi committed:
more solidity style guide changes

koirikivi committed:
Use contract balances for arbitrage and liquidation

Also make it so only executor can call

koirikivi committed:
Use solidity 0.8 compatible openzeppelin in integration tests

koirikivi committed:
Fix liquidator issue in liquidator v2 too

koirikivi committed:
Merge branch 'development' into v2

koirikivi committed:
Arbitrage V2 backend

koirikivi committed:
Refactor ArbitrageV2 to extend Arbitrage v1

Doing so also includes all recent changes to Arbitrage v1

koirikivi committed:
Handle RBTC transfers from LoanClosings

koirikivi committed:
Liquidator V2 backend

koirikivi committed:
Refactor LiquidatorV2 to extend Liquidator v1

koirikivi committed:
Control enabled V2 services from config

koirikivi committed:
DIY Hardhat deployment

koirikivi committed:
Hardhat task for managing Watcher roles

koirikivi committed:
re-tabulate .ts files with 4 spaces and add editorconfig

koirikivi committed:
Separate hardhat tasks into its own file

koirikivi committed:
Watcher funding task

koirikivi committed:
Watcher V2 readme

koirikivi committed:
small fix

koirikivi committed:
Fix gas limit for RBTC withdrawal

koirikivi committed:
Refactoring

koirikivi committed:
Check balances in funding script

+ minor improvement, show WRBTC balance instead of RBTC balance for the
Watcher

koirikivi committed:
mention checking balances in readme

koirikivi committed:
Show watcher contract balances on UI

koirikivi committed:
Make sovrynProtocol the first argument to Watcher

Also fixes a stupid bug in the hardhat deployment script...

koirikivi committed:
re-tab Watcher.sol

koirikivi committed:
Add maintenance methods for updating addresses to Watcher

koirikivi committed:
Update hardhat and use supported Solidity

koirikivi committed:
Remove useless console.log

koirikivi committed:
log v2 usage on startup

koirikivi committed:
Fix small issue in README

koirikivi committed:
Improve Watcher.checkArbitrage

koirikivi committed:
Fix/test Watcher checkArbitrage bisecting

koirikivi committed:
Optimize checkArbitrage by not querying price feed return every time

koirikivi committed:
Watcher cleanup + remove/comment out console.logs

koirikivi committed:
Minor comparison thing

koirikivi committed:
Remove checkArbitrage from the PR as it's not used by backend

koirikivi committed:
Remove whatever this is

koirikivi committed:
Remove explicit return from Watcher.liquidate

koirikivi committed:
Optional swapback for watcher v2

korepkorep commented on smart-contracts/contracts/Watcher.sol:
```diff
@@ -0,0 +1,313 @@
+            wrbtcToken.withdraw(_amount);
+            _receiver.transfer(_amount);
+        } else {
+            _token.transfer(_receiver, _amount);
```
doesn't check returned value

koirikivi replied:
Quick question: Would it be better to use OpenZeppelin SafeERC20 for all ERC20 interactions?

korepkorep commented on smart-contracts/contracts/Watcher.sol:
```diff
@@ -0,0 +1,313 @@
+            require(msg.value == _amount, "Watcher: _amount and msg.value must match for RBTC deposits");
+            wrbtcToken.deposit{ value: _amount }();
+        } else {
+            _token.transferFrom(msg.sender, address(this), _amount);
```
Returned value is not checked

korepkorep commented on smart-contracts/contracts/Watcher.sol:
```diff
@@ -0,0 +1,313 @@
+        (IERC20 sourceToken, IERC20 targetToken, uint256 targetTokenAmount) = swapInternal(
+            _conversionPath,
+            _amount,
+            1
```
It doesn't protect from slippage.  `minReturn` should be passed as parameter of  `arbitrage` or calculated using user values

korepkorep commented on smart-contracts/contracts/Watcher.sol:
```diff
@@ -0,0 +1,313 @@
+            (IERC20 sourceToken, IERC20 targetToken, uint256 targetTokenAmount) = swapInternal(
+                _swapbackConversionPath,
+                _seizedAmount,
+                1
```
It doesn't protect from slippage. `minReturn` should be passed as parameter of  `swapbackInternal` or calculated using other user-provided data

korepkorep commented on smart-contracts/contracts/Watcher.sol:
```diff
@@ -0,0 +1,313 @@
+        );
+
+        uint256 priceFeedReturn = priceFeeds.queryReturn(address(sourceToken), address(targetToken), _amount);
+        uint256 profit = targetTokenAmount - priceFeedReturn;
+        require(profit >= _minProfit, "Watcher: minimum profit not met");
```
In case of underflow (`priceFeedReturn` > `targetTokenAmount`) `require` is pointless since means `+inf > _minProfit` which is always true

koirikivi replied:
Since the contract requires solidity 0.8, it should revert on underflow. However underflow will still revert the call which is probably not ideal.

korepkorep commented on smart-contracts/contracts/Watcher.sol:
```diff
@@ -0,0 +1,313 @@
+    external
+    onlyRole(ROLE_EXECUTOR)
+    {
+        (IERC20 sourceToken, IERC20 targetToken, uint256 targetTokenAmount) = swapInternal(
+            _conversionPath,
+            _amount,
+            1
+        );
+
+        uint256 priceFeedReturn = priceFeeds.queryReturn(address(sourceToken), address(targetToken), _amount);
+        uint256 profit = targetTokenAmount - priceFeedReturn;
+        require(profit >= _minProfit, "Watcher: minimum profit not met");
```
Arbitrage makes sense when the exchange chain is closed (for example, A -> B -> C -> A), otherwise you take additional risks (e.g. what if C price drops?). 
For closed paths this method can be replaced by a call to `swapInternal`, where `_minReturn` is `_amount` + `_minProfit`. It will also help avoid underflow. However, current implementation of `swapInternal` includes a check that the last and first tokens are different

koirikivi replied:
Replacing this check with `_minReturn` change is a good point -- will do the changes as requested.

"Arbitrage" in the usage of this contract is maybe not 100% accurate, since just means buying or selling a token at a price that's better than the price from an external source (price feed). In actual arbitrage, you would need to do the opposite action (sell/buy) on another exchange, or complete a loop and end up with the same asset (as suggested -- perhaps the require check is unnecessary and avoids some actual use cases, though I'm doubtful it can actually happen because internally all assets are traded against WRBTC).

However this implementation is consistent with the way arbitrage works in sovryn-node v1.

korepkorep reviewed (commented):
Several code quality issues that don't endanger contracts` security

koirikivi committed:
Remove unfortunate it.only from integration tests

koirikivi committed:
Code review changes

- Use minReturn instead of doing requires ourselves
- Refactor source/target token splitting out of swapInternal

koirikivi committed:
Refactor swapback so it uses less gas if swapback not performed

koirikivi committed:
Add SPDX license to Watcher.sol

koirikivi committed:
Use SafeERC20 for ERC20 interactions

koirikivi commented:
@korepkorep / @eMarchenko : the findings should be fixed now -- please review and tell me if there's still something.

Genysys committed:
chore: add pm2 to package.json

Genysys committed:
chore: update README.md to reflect pm2

Genysys committed:
fix: updates XUSD address

Genysys committed:
Merge branch 'v2' into chore/pm2

Genysys committed:
fix: add additonal flags to pm2 command

Genysys committed:
Merge branch 'chore/pm2' of https://github.com/DistributedCollective/Sovryn-Node into chore/pm2

Genysys committed:
chore: add npm scripts to run nodemon locally

koirikivi committed:
Merge pull request #115 from DistributedCollective/chore/pm2

Chore/pm2

koirikivi committed:
Merge branch 'development' into v2-merge

koirikivi committed:
Watcher V2 contract: Allow everyone to send RBTC (wrap to WRBTC)

koirikivi committed:
Watcher V2 contract: simplify WRBTC wrapping

koirikivi committed:
Typo, should be RSK and not rsk-testnet here

koirikivi committed:
Add encode-withdrawal task for withdrawing with multisig

koirikivi committed:
log errors with console.error

koirikivi committed:
Fix this liquidator profit thing

koirikivi committed:
Update sovryn-smart-contracts used in integration-tests

koirikivi committed:
Watcher V2: Add tests for basic liquidations

koirikivi committed:
Fix mess-up in tests and improve TestLoanProtocol contract

koirikivi committed:
More liquidations/swapback tests

koirikivi committed:
Watcher V2: Fix bug in liquidation if maxLiquidatable < closeAmount

koirikivi committed:
Watcher V2: prevent leftover allowance by not allowing closeAmount > maxLiquidatable

koirikivi committed:
Check that loan has principal before rollover

koirikivi committed:
Check that position is healthy before rollover

koirikivi committed:
Try simulating rollover transaction before sending

eMarchenko reviewed (approved):
looks good

koirikivi committed:
Update testnet swap address

koirikivi committed:
Method to update sovryn-swap-network

koirikivi committed:
Fix liquidation balance errors hopefully

koirikivi committed:
Try to fix liq errors again

koirikivi committed:
Fix

koirikivi committed:
Fixxx

koirikivi committed:
Revert if swapback doesn't happen

koirikivi committed:
Don't revert on non-profitable swapback after all

koirikivi committed:
Revert "Update testnet swap address"

This reverts commit e10458d767e624f4b36238529d5e993cf41159f4.

koirikivi committed:
Quick and dirty bignumber fix

koirikivi committed:
Another bignumber fix

koirikivi committed:
Liquidate loans with most to liquidate first

Approximases highest value first.

koirikivi committed:
Improve liquidations by value a bit

koirikivi committed:
Fix loadActivePositions (second parameter is nowadays actually count)
